File: problems/0009-palindrome-number/solution.js
/**
 * @problemNumber: 9
 * @problemName: Palindrome Number
 * @difficulty: Easy
 * @topics: Math
 * @companies: Facebook, Apple, Google
 */

/**
 * Approach 1: Revert Half of the Number
 * Time Complexity: O(log10(n))
 * Space Complexity: O(1)
 *
 * @param {number} x
 * @return {boolean}
 */
function isPalindrome(x) {
    // Special cases:
    // x < 0: Negative numbers are not palindromes
    // x % 10 == 0 && x != 0: Numbers ending in 0 are not palindromes (except 0)
    if (x < 0 || (x % 10 === 0 && x !== 0)) {
        return false;
    }

    let revertedNumber = 0;
    while (x > revertedNumber) {
        revertedNumber = revertedNumber * 10 + (x % 10);
        x = Math.floor(x / 10);
    }

    // When the length is an odd number, we can get rid of the middle digit by revertedNumber/10
    // Example: x = 12321, at the end x = 12, revertedNumber = 123
    return x === revertedNumber || x === Math.floor(revertedNumber / 10);
}

// Test cases
console.log(isPalindrome(121));  // true
console.log(isPalindrome(-121)); // false
console.log(isPalindrome(10));   // false

module.exports = { isPalindrome };

File: problems/0009-palindrome-number/solution.py
"""
Problem Number: 9
Problem Name: Palindrome Number
Difficulty: Easy
Topics: Math
Companies: Facebook, Apple, Google
"""

class Solution:
    def isPalindrome(self, x: int) -> bool:
        """
        Approach: Revert Half of the Number
        Time Complexity: O(log10(n))
        Space Complexity: O(1)
        """
        if x < 0 or (x % 10 == 0 and x != 0):
            return False
            
        reverted_number = 0
        while x > reverted_number:
            reverted_number = reverted_number * 10 + x % 10
            x //= 10
            
        return x == reverted_number or x == reverted_number // 10

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.isPalindrome(121))  # True
    print(solution.isPalindrome(-121)) # False
    print(solution.isPalindrome(10))   # False
